


<!DOCTYPE html>
<html id="htmlId">
<head>
  <meta http-equiv="Content-Type" content="text/html;charset=UTF-8"> 
  <title>Coverage Report > BandedLinearAligner</title>
  <style type="text/css">
    @import "../../css/coverage.css";
    @import "../../css/idea.min.css";
  </style>
  <script type="text/javascript" src="../../js/highlight.min.js"></script>
  <script type="text/javascript" src="../../js/highlightjs-line-numbers.min.js"></script>
</head>

<body>
<div class="content">
<div class="breadCrumbs">
Current scope:     <a href="../../index.html">all classes</a>
    <span class="separator">|</span>
    <a href="../index.html">com.milaboratory.core.alignment</a>
</div>

<h1>Coverage Summary for Class: BandedLinearAligner (com.milaboratory.core.alignment)</h1>

<table class="coverageStats">
<tr>
  <th class="name">Class</th>
<th class="coverageStat 
">
  Class, %
</th>
<th class="coverageStat 
">
  Method, %
</th>
<th class="coverageStat 
">
  Line, %
</th>
</tr>
<tr>
  <td class="name">BandedLinearAligner</td>
<td class="coverageStat">
  <span class="percent">
    0%
  </span>
  <span class="absValue">
    (0/1)
  </span>
</td>
<td class="coverageStat">
  <span class="percent">
    0%
  </span>
  <span class="absValue">
    (0/13)
  </span>
</td>
<td class="coverageStat">
  <span class="percent">
    0%
  </span>
  <span class="absValue">
    (0/306)
  </span>
</td>
</tr>

</table>

<br/>
<br/>


<pre>
<code class="sourceCode" id="sourceCode">&nbsp;/*
&nbsp; * Copyright 2015 MiLaboratory.com
&nbsp; *
&nbsp; * Licensed under the Apache License, Version 2.0 (the &quot;License&quot;);
&nbsp; * you may not use this file except in compliance with the License.
&nbsp; * You may obtain a copy of the License at
&nbsp; *
&nbsp; *     http://www.apache.org/licenses/LICENSE-2.0
&nbsp; *
&nbsp; * Unless required by applicable law or agreed to in writing, software
&nbsp; * distributed under the License is distributed on an &quot;AS IS&quot; BASIS,
&nbsp; * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
&nbsp; * See the License for the specific language governing permissions and
&nbsp; * limitations under the License.
&nbsp; */
&nbsp;package com.milaboratory.core.alignment;
&nbsp;
&nbsp;import com.milaboratory.core.Range;
&nbsp;import com.milaboratory.core.mutations.MutationsBuilder;
&nbsp;import com.milaboratory.core.sequence.NucleotideSequence;
&nbsp;
&nbsp;public final class BandedLinearAligner {
&nbsp;    private BandedLinearAligner() {
&nbsp;    }
&nbsp;
&nbsp;    /**
&nbsp;     * Classical Banded Alignment
&nbsp;     * &lt;p/&gt;
&nbsp;     * &lt;p&gt;Both sequences must be highly similar&lt;/p&gt; &lt;p&gt;Align 2 sequence completely (i.e. while first sequence will be
&nbsp;     * aligned against whole second sequence)&lt;/p&gt;
&nbsp;     *
&nbsp;     * @param scoring     scoring system
&nbsp;     * @param seq1        first sequence
&nbsp;     * @param seq2        second sequence
&nbsp;     * @param offset1     offset in first sequence
&nbsp;     * @param length1     length of first sequence&#39;s part to be aligned
&nbsp;     * @param offset2     offset in second sequence
&nbsp;     * @param length2     length of second sequence&#39;s part to be aligned
&nbsp;     * @param width       width of banded alignment matrix. In other terms max allowed number of indels
&nbsp;     * @param mutations   mutations array where all mutations will be kept
&nbsp;     * @param cachedArray cached (created once) array to be used in {@link BandedMatrix}, which is compact alignment
&nbsp;     *                    scoring matrix
&nbsp;     */
&nbsp;    public static float align0(LinearGapAlignmentScoring scoring, NucleotideSequence seq1, NucleotideSequence seq2,
&nbsp;                               int offset1, int length1, int offset2, int length2,
&nbsp;                               int width, MutationsBuilder&lt;NucleotideSequence&gt; mutations, CachedIntArray cachedArray) {
<b class="nc">&nbsp;        if(offset1 &lt; 0 || length1 &lt; 0 || offset2 &lt; 0 || length2 &lt; 0)</b>
<b class="nc">&nbsp;            throw new IllegalArgumentException();</b>
&nbsp;
<b class="nc">&nbsp;        int size1 = length1 + 1,</b>
<b class="nc">&nbsp;                size2 = length2 + 1;</b>
&nbsp;
<b class="nc">&nbsp;        BandedMatrix matrix = new BandedMatrix(cachedArray, size1, size2, width);</b>
&nbsp;
&nbsp;        int i, j;
&nbsp;
<b class="nc">&nbsp;        for (i = matrix.getRowFactor() - matrix.getColumnDelta(); i &gt; 0; --i)</b>
<b class="nc">&nbsp;            matrix.set(0, i, scoring.getGapPenalty() * i);</b>
&nbsp;
<b class="nc">&nbsp;        for (i = matrix.getColumnDelta(); i &gt; 0; --i)</b>
<b class="nc">&nbsp;            matrix.set(i, 0, scoring.getGapPenalty() * i);</b>
&nbsp;
<b class="nc">&nbsp;        matrix.set(0, 0, 0);</b>
&nbsp;
&nbsp;        int match, delete, insert, to;
&nbsp;
<b class="nc">&nbsp;        for (i = 0; i &lt; length1; ++i) {</b>
<b class="nc">&nbsp;            to = Math.min(i + matrix.getRowFactor() - matrix.getColumnDelta() + 1, length2);</b>
<b class="nc">&nbsp;            for (j = Math.max(0, i - matrix.getColumnDelta()); j &lt; to; ++j) {</b>
<b class="nc">&nbsp;                match = matrix.get(i, j) +</b>
<b class="nc">&nbsp;                        scoring.getScore(seq1.codeAt(offset1 + i), seq2.codeAt(offset2 + j));</b>
<b class="nc">&nbsp;                delete = matrix.get(i, j + 1) + scoring.getGapPenalty();</b>
<b class="nc">&nbsp;                insert = matrix.get(i + 1, j) + scoring.getGapPenalty();</b>
<b class="nc">&nbsp;                matrix.set(i + 1, j + 1, Math.max(match, Math.max(delete, insert)));</b>
&nbsp;            }
&nbsp;        }
&nbsp;
<b class="nc">&nbsp;        to = mutations.size();</b>
<b class="nc">&nbsp;        i = length1 - 1;</b>
<b class="nc">&nbsp;        j = length2 - 1;</b>
&nbsp;        byte c1, c2;
<b class="nc">&nbsp;        while (i &gt;= 0 || j &gt;= 0) {</b>
<b class="nc">&nbsp;            if (i &gt;= 0 &amp;&amp; j &gt;= 0 &amp;&amp;</b>
<b class="nc">&nbsp;                    matrix.get(i + 1, j + 1) == matrix.get(i, j) +</b>
<b class="nc">&nbsp;                            scoring.getScore(c1 = seq1.codeAt(offset1 + i),</b>
<b class="nc">&nbsp;                                    c2 = seq2.codeAt(offset2 + j))) {</b>
<b class="nc">&nbsp;                if (c1 != c2)</b>
<b class="nc">&nbsp;                    mutations.appendSubstitution(offset1 + i, c1, c2);</b>
<b class="nc">&nbsp;                --i;</b>
<b class="nc">&nbsp;                --j;</b>
<b class="nc">&nbsp;            } else if (i &gt;= 0 &amp;&amp;</b>
<b class="nc">&nbsp;                    matrix.get(i + 1, j + 1) ==</b>
<b class="nc">&nbsp;                            matrix.get(i, j + 1) + scoring.getGapPenalty()) {</b>
<b class="nc">&nbsp;                mutations.appendDeletion(offset1 + i, seq1.codeAt(offset1 + i));</b>
<b class="nc">&nbsp;                --i;</b>
<b class="nc">&nbsp;            } else if (j &gt;= 0 &amp;&amp;</b>
<b class="nc">&nbsp;                    matrix.get(i + 1, j + 1) ==</b>
<b class="nc">&nbsp;                            matrix.get(i + 1, j) + scoring.getGapPenalty()) {</b>
<b class="nc">&nbsp;                mutations.appendInsertion(offset1 + i + 1, seq2.codeAt(offset2 + j));</b>
<b class="nc">&nbsp;                --j;</b>
&nbsp;            } else
<b class="nc">&nbsp;                throw new RuntimeException();</b>
&nbsp;        }
&nbsp;
<b class="nc">&nbsp;        mutations.reverseRange(to, mutations.size());</b>
<b class="nc">&nbsp;        return matrix.get(length1, length2);</b>
&nbsp;    }
&nbsp;
&nbsp;    /**
&nbsp;     * Semi-semi-global alignment with artificially added letters.
&nbsp;     * &lt;p/&gt;
&nbsp;     * &lt;p&gt;Alignment where second sequence is aligned to the right part of first sequence.&lt;/p&gt; &lt;p&gt;Whole second sequence
&nbsp;     * must be highly similar to the first sequence&lt;/p&gt;
&nbsp;     *
&nbsp;     * @param scoring           scoring system
&nbsp;     * @param seq1              first sequence
&nbsp;     * @param seq2              second sequence
&nbsp;     * @param offset1           offset in first sequence
&nbsp;     * @param length1           length of first sequence&#39;s part to be aligned including artificially added letters
&nbsp;     * @param addedNucleotides1 number of artificially added letters to the first sequence
&nbsp;     * @param offset2           offset in second sequence
&nbsp;     * @param length2           length of second sequence&#39;s part to be aligned including artificially added letters
&nbsp;     * @param addedNucleotides2 number of artificially added letters to the second sequence
&nbsp;     * @param width             width of banded alignment matrix. In other terms max allowed number of indels
&nbsp;     * @param mutations         mutations array where all mutations will be kept
&nbsp;     * @param cachedArray       cached (created once) array to be used in {@link BandedMatrix}, which is compact
&nbsp;     *                          alignment scoring matrix
&nbsp;     */
&nbsp;    public static BandedSemiLocalResult alignRightAdded0(LinearGapAlignmentScoring scoring, NucleotideSequence seq1, NucleotideSequence seq2,
&nbsp;                                                         int offset1, int length1, int addedNucleotides1, int offset2, int length2, int addedNucleotides2,
&nbsp;                                                         int width, MutationsBuilder&lt;NucleotideSequence&gt; mutations, CachedIntArray cachedArray) {
<b class="nc">&nbsp;        if(offset1 &lt; 0 || length1 &lt; 0 || offset2 &lt; 0 || length2 &lt; 0)</b>
<b class="nc">&nbsp;            throw new IllegalArgumentException();</b>
&nbsp;
<b class="nc">&nbsp;        int size1 = length1 + 1,</b>
<b class="nc">&nbsp;                size2 = length2 + 1;</b>
&nbsp;
<b class="nc">&nbsp;        BandedMatrix matrix = new BandedMatrix(cachedArray, size1, size2, width);</b>
&nbsp;
&nbsp;        int i, j;
&nbsp;
<b class="nc">&nbsp;        for (i = matrix.getRowFactor() - matrix.getColumnDelta(); i &gt; 0; --i)</b>
<b class="nc">&nbsp;            matrix.set(0, i, scoring.getGapPenalty() * i);</b>
&nbsp;
<b class="nc">&nbsp;        for (i = matrix.getColumnDelta(); i &gt; 0; --i)</b>
<b class="nc">&nbsp;            matrix.set(i, 0, scoring.getGapPenalty() * i);</b>
&nbsp;
<b class="nc">&nbsp;        matrix.set(0, 0, 0);</b>
&nbsp;
&nbsp;        int match, delete, insert, to;
&nbsp;
<b class="nc">&nbsp;        for (i = 0; i &lt; length1; ++i) {</b>
<b class="nc">&nbsp;            to = Math.min(i + matrix.getRowFactor() - matrix.getColumnDelta() + 1, length2);</b>
<b class="nc">&nbsp;            for (j = Math.max(0, i - matrix.getColumnDelta()); j &lt; to; ++j) {</b>
<b class="nc">&nbsp;                match = matrix.get(i, j) +</b>
<b class="nc">&nbsp;                        scoring.getScore(seq1.codeAt(offset1 + i), seq2.codeAt(offset2 + j));</b>
<b class="nc">&nbsp;                delete = matrix.get(i, j + 1) + scoring.getGapPenalty();</b>
<b class="nc">&nbsp;                insert = matrix.get(i + 1, j) + scoring.getGapPenalty();</b>
<b class="nc">&nbsp;                matrix.set(i + 1, j + 1, Math.max(match, Math.max(delete, insert)));</b>
&nbsp;            }
&nbsp;        }
&nbsp;
&nbsp;        //Searching for max.
<b class="nc">&nbsp;        int maxI = -1, maxJ = -1;</b>
<b class="nc">&nbsp;        int maxScore = Integer.MIN_VALUE;</b>
&nbsp;
<b class="nc">&nbsp;        j = length2;</b>
<b class="nc">&nbsp;        for (i = length1 - addedNucleotides1; i &lt; size1; ++i)</b>
<b class="nc">&nbsp;            if (maxScore &lt; matrix.get(i, j)) {</b>
<b class="nc">&nbsp;                maxScore = matrix.get(i, j);</b>
<b class="nc">&nbsp;                maxI = i;</b>
<b class="nc">&nbsp;                maxJ = j;</b>
&nbsp;            }
&nbsp;
<b class="nc">&nbsp;        i = length1;</b>
<b class="nc">&nbsp;        for (j = length2 - addedNucleotides2; j &lt; size2; ++j)</b>
<b class="nc">&nbsp;            if (maxScore &lt; matrix.get(i, j)) {</b>
<b class="nc">&nbsp;                maxScore = matrix.get(i, j);</b>
<b class="nc">&nbsp;                maxI = i;</b>
<b class="nc">&nbsp;                maxJ = j;</b>
&nbsp;            }
&nbsp;
<b class="nc">&nbsp;        to = mutations.size();</b>
<b class="nc">&nbsp;        i = maxI - 1;</b>
<b class="nc">&nbsp;        j = maxJ - 1;</b>
&nbsp;        byte c1, c2;
<b class="nc">&nbsp;        while (i &gt;= 0 || j &gt;= 0) {</b>
<b class="nc">&nbsp;            if (i &gt;= 0 &amp;&amp; j &gt;= 0 &amp;&amp;</b>
<b class="nc">&nbsp;                    matrix.get(i + 1, j + 1) == matrix.get(i, j) +</b>
<b class="nc">&nbsp;                            scoring.getScore(c1 = seq1.codeAt(offset1 + i),</b>
<b class="nc">&nbsp;                                    c2 = seq2.codeAt(offset2 + j))) {</b>
<b class="nc">&nbsp;                if (c1 != c2)</b>
<b class="nc">&nbsp;                    mutations.appendSubstitution(offset1 + i, c1, c2);</b>
<b class="nc">&nbsp;                --i;</b>
<b class="nc">&nbsp;                --j;</b>
<b class="nc">&nbsp;            } else if (i &gt;= 0 &amp;&amp;</b>
<b class="nc">&nbsp;                    matrix.get(i + 1, j + 1) ==</b>
<b class="nc">&nbsp;                            matrix.get(i, j + 1) + scoring.getGapPenalty()) {</b>
<b class="nc">&nbsp;                mutations.appendDeletion(offset1 + i, seq1.codeAt(offset1 + i));</b>
<b class="nc">&nbsp;                --i;</b>
<b class="nc">&nbsp;            } else if (j &gt;= 0 &amp;&amp;</b>
<b class="nc">&nbsp;                    matrix.get(i + 1, j + 1) ==</b>
<b class="nc">&nbsp;                            matrix.get(i + 1, j) + scoring.getGapPenalty()) {</b>
<b class="nc">&nbsp;                mutations.appendInsertion(offset1 + i + 1, seq2.codeAt(offset2 + j));</b>
<b class="nc">&nbsp;                --j;</b>
&nbsp;            } else
<b class="nc">&nbsp;                throw new RuntimeException();</b>
&nbsp;        }
&nbsp;
<b class="nc">&nbsp;        mutations.reverseRange(to, mutations.size());</b>
&nbsp;
<b class="nc">&nbsp;        return new BandedSemiLocalResult(offset1 + maxI - 1, offset2 + maxJ - 1, maxScore);</b>
&nbsp;    }
&nbsp;
&nbsp;
&nbsp;    /**
&nbsp;     * Semi-semi-global alignment with artificially added letters.
&nbsp;     *
&nbsp;     * &lt;p&gt;Alignment where second sequence is aligned to the left part of first sequence.&lt;/p&gt;
&nbsp;     *
&nbsp;     * &lt;p&gt;Whole second sequence must be highly similar to the first sequence, except last {@code width} letters, which
&nbsp;     * are to be checked whether they can improve alignment or not.&lt;/p&gt;
&nbsp;     *
&nbsp;     * @param scoring           scoring system
&nbsp;     * @param seq1              first sequence
&nbsp;     * @param seq2              second sequence
&nbsp;     * @param offset1           offset in first sequence
&nbsp;     * @param length1           length of first sequence&#39;s part to be aligned
&nbsp;     * @param addedNucleotides1 number of artificially added letters to the first sequence;
&nbsp;     *                          must be 0 if seq1 is a pattern to match and seq2 is target sequence
&nbsp;     *                          where we search the pattern
&nbsp;     * @param offset2           offset in second sequence
&nbsp;     * @param length2           length of second sequence&#39;s part to be aligned
&nbsp;     * @param addedNucleotides2 number of artificially added letters to the second sequence;
&nbsp;     *                          if seq2 is target sequence where we search the pattern, this parameter must be equal
&nbsp;     *                          to maximum allowed number of indels (same as width parameter)
&nbsp;     * @param width             width of banded alignment matrix. In other terms max allowed number of indels
&nbsp;     * @param mutations         mutations array where all mutations will be kept
&nbsp;     * @param cachedArray       cached (created once) array to be used in {@link BandedMatrix}, which is compact
&nbsp;     *                          alignment scoring matrix
&nbsp;     */
&nbsp;    public static BandedSemiLocalResult alignLeftAdded0(LinearGapAlignmentScoring scoring, NucleotideSequence seq1, NucleotideSequence seq2,
&nbsp;                                                        int offset1, int length1, int addedNucleotides1, int offset2, int length2, int addedNucleotides2,
&nbsp;                                                        int width, MutationsBuilder&lt;NucleotideSequence&gt; mutations, CachedIntArray cachedArray) {
<b class="nc">&nbsp;        if(offset1 &lt; 0 || length1 &lt; 0 || offset2 &lt; 0 || length2 &lt; 0)</b>
<b class="nc">&nbsp;            throw new IllegalArgumentException();</b>
&nbsp;
<b class="nc">&nbsp;        int size1 = length1 + 1,</b>
<b class="nc">&nbsp;                size2 = length2 + 1;</b>
&nbsp;
<b class="nc">&nbsp;        BandedMatrix matrix = new BandedMatrix(cachedArray, size1, size2, width);</b>
&nbsp;
&nbsp;        int i, j;
&nbsp;
<b class="nc">&nbsp;        for (i = matrix.getRowFactor() - matrix.getColumnDelta(); i &gt; 0; --i)</b>
<b class="nc">&nbsp;            matrix.set(0, i, scoring.getGapPenalty() * i);</b>
&nbsp;
<b class="nc">&nbsp;        for (i = matrix.getColumnDelta(); i &gt; 0; --i)</b>
<b class="nc">&nbsp;            matrix.set(i, 0, scoring.getGapPenalty() * i);</b>
&nbsp;
<b class="nc">&nbsp;        matrix.set(0, 0, 0);</b>
&nbsp;
&nbsp;        int match, delete, insert, to;
&nbsp;
<b class="nc">&nbsp;        for (i = 0; i &lt; length1; ++i) {</b>
<b class="nc">&nbsp;            to = Math.min(i + matrix.getRowFactor() - matrix.getColumnDelta() + 1, length2);</b>
<b class="nc">&nbsp;            for (j = Math.max(0, i - matrix.getColumnDelta()); j &lt; to; ++j) {</b>
<b class="nc">&nbsp;                match = matrix.get(i, j) +</b>
<b class="nc">&nbsp;                        scoring.getScore(seq1.codeAt(offset1 + length1 - 1 - i),</b>
<b class="nc">&nbsp;                                seq2.codeAt(offset2 + length2 - 1 - j));</b>
<b class="nc">&nbsp;                delete = matrix.get(i, j + 1) + scoring.getGapPenalty();</b>
<b class="nc">&nbsp;                insert = matrix.get(i + 1, j) + scoring.getGapPenalty();</b>
<b class="nc">&nbsp;                matrix.set(i + 1, j + 1, Math.max(match, Math.max(delete, insert)));</b>
&nbsp;            }
&nbsp;        }
&nbsp;
&nbsp;        //Searching for max.
<b class="nc">&nbsp;        int maxI = -1, maxJ = -1;</b>
<b class="nc">&nbsp;        int maxScore = Integer.MIN_VALUE;</b>
&nbsp;
<b class="nc">&nbsp;        j = length2;</b>
<b class="nc">&nbsp;        for (i = length1 - addedNucleotides1; i &lt; size1; ++i)</b>
<b class="nc">&nbsp;            if (maxScore &lt; matrix.get(i, j)) {</b>
<b class="nc">&nbsp;                maxScore = matrix.get(i, j);</b>
<b class="nc">&nbsp;                maxI = i;</b>
<b class="nc">&nbsp;                maxJ = j;</b>
&nbsp;            }
&nbsp;
<b class="nc">&nbsp;        i = length1;</b>
<b class="nc">&nbsp;        for (j = length2 - addedNucleotides2; j &lt; size2; ++j)</b>
<b class="nc">&nbsp;            if (maxScore &lt; matrix.get(i, j)) {</b>
<b class="nc">&nbsp;                maxScore = matrix.get(i, j);</b>
<b class="nc">&nbsp;                maxI = i;</b>
<b class="nc">&nbsp;                maxJ = j;</b>
&nbsp;            }
&nbsp;
<b class="nc">&nbsp;        i = maxI - 1;</b>
<b class="nc">&nbsp;        j = maxJ - 1;</b>
&nbsp;        byte c1, c2;
<b class="nc">&nbsp;        while (i &gt;= 0 || j &gt;= 0) {</b>
<b class="nc">&nbsp;            if (i &gt;= 0 &amp;&amp; j &gt;= 0 &amp;&amp;</b>
<b class="nc">&nbsp;                    matrix.get(i + 1, j + 1) == matrix.get(i, j) +</b>
<b class="nc">&nbsp;                            scoring.getScore(c1 = seq1.codeAt(offset1 + length1 - 1 - i),</b>
<b class="nc">&nbsp;                                    c2 = seq2.codeAt(offset2 + length2 - 1 - j))) {</b>
<b class="nc">&nbsp;                if (c1 != c2)</b>
<b class="nc">&nbsp;                    mutations.appendSubstitution(offset1 + length1 - 1 - i, c1, c2);</b>
<b class="nc">&nbsp;                --i;</b>
<b class="nc">&nbsp;                --j;</b>
<b class="nc">&nbsp;            } else if (i &gt;= 0 &amp;&amp;</b>
<b class="nc">&nbsp;                    matrix.get(i + 1, j + 1) ==</b>
<b class="nc">&nbsp;                            matrix.get(i, j + 1) + scoring.getGapPenalty()) {</b>
<b class="nc">&nbsp;                mutations.appendDeletion(offset1 + length1 - 1 - i, seq1.codeAt(offset1 + length1 - 1 - i));</b>
<b class="nc">&nbsp;                --i;</b>
<b class="nc">&nbsp;            } else if (j &gt;= 0 &amp;&amp;</b>
<b class="nc">&nbsp;                    matrix.get(i + 1, j + 1) ==</b>
<b class="nc">&nbsp;                            matrix.get(i + 1, j) + scoring.getGapPenalty()) {</b>
<b class="nc">&nbsp;                mutations.appendInsertion(offset1 + length1 - 1 - i, seq2.codeAt(offset2 + length2 - 1 - j));</b>
<b class="nc">&nbsp;                --j;</b>
&nbsp;            } else
<b class="nc">&nbsp;                throw new RuntimeException();</b>
&nbsp;        }
&nbsp;
<b class="nc">&nbsp;        return new BandedSemiLocalResult(offset1 + length1 - maxI, offset2 + length2 - maxJ, maxScore);</b>
&nbsp;    }
&nbsp;
&nbsp;
&nbsp;    /**
&nbsp;     * Alignment which identifies what is the highly similar part of the both sequences.
&nbsp;     * &lt;p/&gt;
&nbsp;     * &lt;p&gt;Alignment is done in the way that beginning of second sequences is aligned to beginning of first
&nbsp;     * sequence.&lt;/p&gt;
&nbsp;     * &lt;p/&gt;
&nbsp;     * &lt;p&gt;Alignment terminates when score in banded alignment matrix reaches {@code stopPenalty} value.&lt;/p&gt;
&nbsp;     * &lt;p/&gt;
&nbsp;     * &lt;p&gt;In other words, only left part of second sequence is to be aligned&lt;/p&gt;
&nbsp;     *
&nbsp;     * @param scoring     scoring system
&nbsp;     * @param seq1        first sequence
&nbsp;     * @param seq2        second sequence
&nbsp;     * @param offset1     offset in first sequence
&nbsp;     * @param length1     length of first sequence&#39;s part to be aligned
&nbsp;     * @param offset2     offset in second sequence
&nbsp;     * @param length2     length of second sequence&#39;s part to be aligned@param width
&nbsp;     * @param stopPenalty alignment score value in banded alignment matrix at which alignment terminates
&nbsp;     * @param mutations   array where all mutations will be kept
&nbsp;     * @param cachedArray cached (created once) array to be used in {@link BandedMatrix}, which is compact alignment
&nbsp;     *                    scoring matrix
&nbsp;     * @return object which contains positions at which alignment terminated and array of mutations
&nbsp;     */
&nbsp;    public static BandedSemiLocalResult alignSemiLocalLeft0(LinearGapAlignmentScoring scoring, NucleotideSequence seq1, NucleotideSequence seq2,
&nbsp;                                                            int offset1, int length1, int offset2, int length2,
&nbsp;                                                            int width, int stopPenalty, MutationsBuilder&lt;NucleotideSequence&gt; mutations,
&nbsp;                                                            CachedIntArray cachedArray) {
<b class="nc">&nbsp;        if(offset1 &lt; 0 || length1 &lt; 0 || offset2 &lt; 0 || length2 &lt; 0)</b>
<b class="nc">&nbsp;            throw new IllegalArgumentException();</b>
&nbsp;
<b class="nc">&nbsp;        int size1 = length1 + 1,</b>
<b class="nc">&nbsp;                size2 = length2 + 1;</b>
&nbsp;
<b class="nc">&nbsp;        int matchReward = scoring.getScore((byte) 0, (byte) 0);</b>
&nbsp;
<b class="nc">&nbsp;        BandedMatrix matrix = new BandedMatrix(cachedArray, size1, size2, width);</b>
&nbsp;
&nbsp;        int i, j;
&nbsp;
<b class="nc">&nbsp;        for (i = matrix.getRowFactor() - matrix.getColumnDelta(); i &gt; 0; --i)</b>
<b class="nc">&nbsp;            matrix.set(0, i, scoring.getGapPenalty() * i);</b>
&nbsp;
<b class="nc">&nbsp;        for (i = matrix.getColumnDelta(); i &gt; 0; --i)</b>
<b class="nc">&nbsp;            matrix.set(i, 0, scoring.getGapPenalty() * i);</b>
&nbsp;
<b class="nc">&nbsp;        matrix.set(0, 0, 0);</b>
&nbsp;
&nbsp;        int match, delete, insert, to;
<b class="nc">&nbsp;        int max = 0;</b>
<b class="nc">&nbsp;        int iStop = 0, jStop = 0;</b>
&nbsp;        int rowMax;
&nbsp;
<b class="nc">&nbsp;        for (i = 0; i &lt; length1; ++i) {</b>
<b class="nc">&nbsp;            to = Math.min(i + matrix.getRowFactor() - matrix.getColumnDelta() + 1, size2 - 1);</b>
<b class="nc">&nbsp;            rowMax = Integer.MIN_VALUE;</b>
<b class="nc">&nbsp;            for (j = Math.max(0, i - matrix.getColumnDelta()); j &lt; to; ++j) {</b>
<b class="nc">&nbsp;                match = matrix.get(i, j) +</b>
<b class="nc">&nbsp;                        scoring.getScore(seq1.codeAt(offset1 + i), seq2.codeAt(offset2 + j));</b>
<b class="nc">&nbsp;                delete = matrix.get(i, j + 1) + scoring.getGapPenalty();</b>
<b class="nc">&nbsp;                insert = matrix.get(i + 1, j) + scoring.getGapPenalty();</b>
<b class="nc">&nbsp;                matrix.set(i + 1, j + 1, match = Math.max(match, Math.max(delete, insert)));</b>
<b class="nc">&nbsp;                if (max &lt; match) {</b>
<b class="nc">&nbsp;                    iStop = i + 1;</b>
<b class="nc">&nbsp;                    jStop = j + 1;</b>
<b class="nc">&nbsp;                    max = match;</b>
&nbsp;                }
<b class="nc">&nbsp;                rowMax = Math.max(rowMax, match);</b>
&nbsp;            }
<b class="nc">&nbsp;            if (rowMax - i * matchReward &lt; stopPenalty)</b>
<b class="nc">&nbsp;                break;</b>
&nbsp;        }
&nbsp;
&nbsp;
<b class="nc">&nbsp;        int fromL = mutations.size();</b>
&nbsp;
<b class="nc">&nbsp;        i = iStop - 1;</b>
<b class="nc">&nbsp;        j = jStop - 1;</b>
&nbsp;        byte c1, c2;
<b class="nc">&nbsp;        while (i &gt;= 0 || j &gt;= 0) {</b>
<b class="nc">&nbsp;            if (i &gt;= 0 &amp;&amp; j &gt;= 0 &amp;&amp;</b>
<b class="nc">&nbsp;                    matrix.get(i + 1, j + 1) == matrix.get(i, j) +</b>
<b class="nc">&nbsp;                            scoring.getScore(c1 = seq1.codeAt(offset1 + i),</b>
<b class="nc">&nbsp;                                    c2 = seq2.codeAt(offset2 + j))) {</b>
<b class="nc">&nbsp;                if (c1 != c2)</b>
<b class="nc">&nbsp;                    mutations.appendSubstitution(offset1 + i, c1, c2);</b>
<b class="nc">&nbsp;                --i;</b>
<b class="nc">&nbsp;                --j;</b>
<b class="nc">&nbsp;            } else if (i &gt;= 0 &amp;&amp;</b>
<b class="nc">&nbsp;                    matrix.get(i + 1, j + 1) ==</b>
<b class="nc">&nbsp;                            matrix.get(i, j + 1) + scoring.getGapPenalty()) {</b>
<b class="nc">&nbsp;                mutations.appendDeletion(offset1 + i, seq1.codeAt(offset1 + i));</b>
<b class="nc">&nbsp;                --i;</b>
<b class="nc">&nbsp;            } else if (j &gt;= 0 &amp;&amp;</b>
<b class="nc">&nbsp;                    matrix.get(i + 1, j + 1) ==</b>
<b class="nc">&nbsp;                            matrix.get(i + 1, j) + scoring.getGapPenalty()) {</b>
<b class="nc">&nbsp;                mutations.appendInsertion(offset1 + i + 1, seq2.codeAt(offset2 + j));</b>
<b class="nc">&nbsp;                --j;</b>
&nbsp;            } else
<b class="nc">&nbsp;                throw new RuntimeException();</b>
&nbsp;        }
&nbsp;
<b class="nc">&nbsp;        mutations.reverseRange(fromL, mutations.size());</b>
&nbsp;
<b class="nc">&nbsp;        return new BandedSemiLocalResult(offset1 + iStop - 1, offset2 + jStop - 1, max);</b>
&nbsp;    }
&nbsp;
&nbsp;    /**
&nbsp;     * Alignment which identifies what is the highly similar part of the both sequences.
&nbsp;     * &lt;p/&gt;
&nbsp;     * &lt;p&gt;Alignment is done in the way that end of second sequence is aligned to end of first sequence.&lt;/p&gt; &lt;p&gt;Alignment
&nbsp;     * terminates when score in banded alignment matrix reaches {@code stopPenalty} value.&lt;/p&gt; &lt;p&gt;In other words, only
&nbsp;     * right part of second sequence is to be aligned.&lt;/p&gt;
&nbsp;     *
&nbsp;     * @param scoring     scoring system
&nbsp;     * @param seq1        first sequence
&nbsp;     * @param seq2        second sequence
&nbsp;     * @param offset1     offset in first sequence
&nbsp;     * @param length1     length of first sequence&#39;s part to be aligned
&nbsp;     * @param offset2     offset in second sequence
&nbsp;     * @param length2     length of second sequence&#39;s part to be aligned@param width
&nbsp;     * @param stopPenalty alignment score value in banded alignment matrix at which alignment terminates
&nbsp;     * @param mutations   array where all mutations will be kept
&nbsp;     * @param cachedArray cached (created once) array to be used in {@link BandedMatrix}, which is compact alignment
&nbsp;     *                    scoring matrix
&nbsp;     * @return object which contains positions at which alignment terminated and array of mutations
&nbsp;     */
&nbsp;    public static BandedSemiLocalResult alignSemiLocalRight0(LinearGapAlignmentScoring scoring, NucleotideSequence seq1, NucleotideSequence seq2,
&nbsp;                                                             int offset1, int length1, int offset2, int length2,
&nbsp;                                                             int width, int stopPenalty, MutationsBuilder&lt;NucleotideSequence&gt; mutations,
&nbsp;                                                             CachedIntArray cachedArray) {
<b class="nc">&nbsp;        if(offset1 &lt; 0 || length1 &lt; 0 || offset2 &lt; 0 || length2 &lt; 0)</b>
<b class="nc">&nbsp;            throw new IllegalArgumentException();</b>
&nbsp;
<b class="nc">&nbsp;        int size1 = length1 + 1,</b>
<b class="nc">&nbsp;                size2 = length2 + 1;</b>
&nbsp;
<b class="nc">&nbsp;        int matchReward = scoring.getScore((byte) 0, (byte) 0);</b>
&nbsp;
<b class="nc">&nbsp;        BandedMatrix matrix = new BandedMatrix(cachedArray, size1, size2, width);</b>
&nbsp;
&nbsp;        int i, j;
&nbsp;
<b class="nc">&nbsp;        for (i = matrix.getRowFactor() - matrix.getColumnDelta(); i &gt; 0; --i)</b>
<b class="nc">&nbsp;            matrix.set(0, i, scoring.getGapPenalty() * i);</b>
&nbsp;
<b class="nc">&nbsp;        for (i = matrix.getColumnDelta(); i &gt; 0; --i)</b>
<b class="nc">&nbsp;            matrix.set(i, 0, scoring.getGapPenalty() * i);</b>
&nbsp;
<b class="nc">&nbsp;        matrix.set(0, 0, 0);</b>
&nbsp;
&nbsp;        int match, delete, insert, to;
<b class="nc">&nbsp;        int max = 0;</b>
<b class="nc">&nbsp;        int iStop = 0, jStop = 0;</b>
&nbsp;        int rowMax;
&nbsp;
<b class="nc">&nbsp;        for (i = 0; i &lt; length1; ++i) {</b>
<b class="nc">&nbsp;            to = Math.min(i + matrix.getRowFactor() - matrix.getColumnDelta() + 1, length2);</b>
<b class="nc">&nbsp;            rowMax = Integer.MIN_VALUE;</b>
<b class="nc">&nbsp;            for (j = Math.max(0, i - matrix.getColumnDelta()); j &lt; to; ++j) {</b>
<b class="nc">&nbsp;                match = matrix.get(i, j) +</b>
<b class="nc">&nbsp;                        scoring.getScore(seq1.codeAt(offset1 + length1 - 1 - i),</b>
<b class="nc">&nbsp;                                seq2.codeAt(offset2 + length2 - 1 - j));</b>
<b class="nc">&nbsp;                delete = matrix.get(i, j + 1) + scoring.getGapPenalty();</b>
<b class="nc">&nbsp;                insert = matrix.get(i + 1, j) + scoring.getGapPenalty();</b>
<b class="nc">&nbsp;                matrix.set(i + 1, j + 1, match = Math.max(match, Math.max(delete, insert)));</b>
<b class="nc">&nbsp;                if (max &lt; match) {</b>
<b class="nc">&nbsp;                    iStop = i + 1;</b>
<b class="nc">&nbsp;                    jStop = j + 1;</b>
<b class="nc">&nbsp;                    max = match;</b>
&nbsp;                }
<b class="nc">&nbsp;                rowMax = Math.max(rowMax, match);</b>
&nbsp;            }
<b class="nc">&nbsp;            if (rowMax - i * matchReward &lt; stopPenalty)</b>
<b class="nc">&nbsp;                break;</b>
&nbsp;        }
&nbsp;
<b class="nc">&nbsp;        i = iStop - 1;</b>
<b class="nc">&nbsp;        j = jStop - 1;</b>
&nbsp;        byte c1, c2;
<b class="nc">&nbsp;        while (i &gt;= 0 || j &gt;= 0) {</b>
<b class="nc">&nbsp;            if (i &gt;= 0 &amp;&amp; j &gt;= 0 &amp;&amp;</b>
<b class="nc">&nbsp;                    matrix.get(i + 1, j + 1) == matrix.get(i, j) +</b>
<b class="nc">&nbsp;                            scoring.getScore(c1 = seq1.codeAt(offset1 + length1 - 1 - i),</b>
<b class="nc">&nbsp;                                    c2 = seq2.codeAt(offset2 + length2 - 1 - j))) {</b>
<b class="nc">&nbsp;                if (c1 != c2)</b>
<b class="nc">&nbsp;                    mutations.appendSubstitution(offset1 + length1 - 1 - i, c1, c2);</b>
<b class="nc">&nbsp;                --i;</b>
<b class="nc">&nbsp;                --j;</b>
<b class="nc">&nbsp;            } else if (i &gt;= 0 &amp;&amp;</b>
<b class="nc">&nbsp;                    matrix.get(i + 1, j + 1) ==</b>
<b class="nc">&nbsp;                            matrix.get(i, j + 1) + scoring.getGapPenalty()) {</b>
<b class="nc">&nbsp;                mutations.appendDeletion(offset1 + length1 - 1 - i, seq1.codeAt(offset1 + length1 - 1 - i));</b>
<b class="nc">&nbsp;                --i;</b>
<b class="nc">&nbsp;            } else if (j &gt;= 0 &amp;&amp;</b>
<b class="nc">&nbsp;                    matrix.get(i + 1, j + 1) ==</b>
<b class="nc">&nbsp;                            matrix.get(i + 1, j) + scoring.getGapPenalty()) {</b>
<b class="nc">&nbsp;                mutations.appendInsertion(offset1 + length1 - 1 - i, seq2.codeAt(offset2 + length2 - 1 - j));</b>
<b class="nc">&nbsp;                --j;</b>
&nbsp;            } else
<b class="nc">&nbsp;                throw new RuntimeException();</b>
&nbsp;        }
&nbsp;
<b class="nc">&nbsp;        return new BandedSemiLocalResult(offset1 + length1 - iStop, offset2 + length2 - jStop, max);</b>
&nbsp;    }
&nbsp;
&nbsp;
&nbsp;    /**
&nbsp;     * Classical Banded Alignment
&nbsp;     * &lt;p/&gt;
&nbsp;     * &lt;p&gt;Both sequences must be highly similar&lt;/p&gt; &lt;p&gt;Align 2 sequence completely (i.e. while first sequence will be
&nbsp;     * aligned against whole second sequence)&lt;/p&gt;
&nbsp;     *
&nbsp;     * @param scoring scoring system
&nbsp;     * @param seq1    first sequence
&nbsp;     * @param seq2    second sequence
&nbsp;     * @param width   width of banded alignment matrix. In other terms max allowed number of indels
&nbsp;     */
&nbsp;    public static Alignment&lt;NucleotideSequence&gt; align(LinearGapAlignmentScoring scoring, NucleotideSequence seq1, NucleotideSequence seq2,
&nbsp;                                                      int width) {
<b class="nc">&nbsp;        return align(scoring, seq1, seq2, 0, seq1.size(), 0, seq2.size(), width);</b>
&nbsp;    }
&nbsp;
&nbsp;    /**
&nbsp;     * Classical Banded Alignment
&nbsp;     * &lt;p/&gt;
&nbsp;     * &lt;p&gt;Both sequences must be highly similar&lt;/p&gt; &lt;p&gt;Align 2 sequence completely (i.e. while first sequence will be
&nbsp;     * aligned against whole second sequence)&lt;/p&gt;
&nbsp;     *
&nbsp;     * @param scoring scoring system
&nbsp;     * @param seq1    first sequence
&nbsp;     * @param seq2    second sequence
&nbsp;     * @param offset1 offset in first sequence
&nbsp;     * @param length1 length of first sequence&#39;s part to be aligned
&nbsp;     * @param offset2 offset in second sequence
&nbsp;     * @param length2 length of second sequence&#39;s part to be aligned
&nbsp;     * @param width   width of banded alignment matrix. In other terms max allowed number of indels
&nbsp;     */
&nbsp;    public static Alignment&lt;NucleotideSequence&gt; align(LinearGapAlignmentScoring scoring, NucleotideSequence seq1, NucleotideSequence seq2,
&nbsp;                                                      int offset1, int length1, int offset2, int length2, int width) {
&nbsp;        try {
<b class="nc">&nbsp;            MutationsBuilder&lt;NucleotideSequence&gt; mutations = new MutationsBuilder&lt;&gt;(NucleotideSequence.ALPHABET);</b>
<b class="nc">&nbsp;            float score = align0(scoring, seq1, seq2, offset1, length1, offset2, length2, width,</b>
<b class="nc">&nbsp;                    mutations, AlignmentCache.get());</b>
<b class="nc">&nbsp;            return new Alignment&lt;&gt;(seq1, mutations.createAndDestroy(),</b>
&nbsp;                    new Range(offset1, offset1 + length1), new Range(offset2, offset2 + length2), score);
&nbsp;        } finally {
<b class="nc">&nbsp;            AlignmentCache.release();</b>
<b class="nc">&nbsp;        }</b>
&nbsp;    }
&nbsp;
&nbsp;    /**
&nbsp;     * Semi-semi-global alignment with artificially added letters.
&nbsp;     *
&nbsp;     * &lt;p&gt;Alignment where second sequence is aligned to the left part of first sequence.&lt;/p&gt;
&nbsp;     *
&nbsp;     * &lt;p&gt;Whole second sequence must be highly similar to the first sequence, except last {@code width} letters, which
&nbsp;     * are to be checked whether they can improve alignment or not.&lt;/p&gt;
&nbsp;     *
&nbsp;     * @param scoring           scoring system
&nbsp;     * @param seq1              first sequence
&nbsp;     * @param seq2              second sequence
&nbsp;     * @param offset1           offset in first sequence
&nbsp;     * @param length1           length of first sequence&#39;s part to be aligned
&nbsp;     * @param addedNucleotides1 number of artificially added letters to the first sequence
&nbsp;     * @param offset2           offset in second sequence
&nbsp;     * @param length2           length of second sequence&#39;s part to be aligned
&nbsp;     * @param addedNucleotides2 number of artificially added letters to the second sequence
&nbsp;     * @param width             width of banded alignment matrix. In other terms max allowed number of indels
&nbsp;     */
&nbsp;    public static Alignment&lt;NucleotideSequence&gt; alignLeftAdded(LinearGapAlignmentScoring scoring, NucleotideSequence seq1, NucleotideSequence seq2,
&nbsp;                                                               int offset1, int length1, int addedNucleotides1, int offset2, int length2, int addedNucleotides2,
&nbsp;                                                               int width) {
&nbsp;        try {
<b class="nc">&nbsp;            MutationsBuilder&lt;NucleotideSequence&gt; mutations = new MutationsBuilder&lt;&gt;(NucleotideSequence.ALPHABET);</b>
<b class="nc">&nbsp;            BandedSemiLocalResult result = alignLeftAdded0(scoring, seq1, seq2,</b>
&nbsp;                    offset1, length1, addedNucleotides1, offset2, length2, addedNucleotides2,
<b class="nc">&nbsp;                    width, mutations, AlignmentCache.get());</b>
<b class="nc">&nbsp;            return new Alignment&lt;&gt;(seq1, mutations.createAndDestroy(),</b>
&nbsp;                    new Range(result.sequence1Stop, offset1 + length1), new Range(result.sequence2Stop, offset2 + length2),
&nbsp;                    result.score);
&nbsp;        } finally {
<b class="nc">&nbsp;            AlignmentCache.release();</b>
<b class="nc">&nbsp;        }</b>
&nbsp;    }
&nbsp;
&nbsp;    /**
&nbsp;     * Semi-semi-global alignment with artificially added letters.
&nbsp;     * &lt;p/&gt;
&nbsp;     * &lt;p&gt;Alignment where second sequence is aligned to the right part of first sequence.&lt;/p&gt; &lt;p&gt;Whole second sequence
&nbsp;     * must be highly similar to the first sequence&lt;/p&gt;
&nbsp;     *
&nbsp;     * @param scoring           scoring system
&nbsp;     * @param seq1              first sequence
&nbsp;     * @param seq2              second sequence
&nbsp;     * @param offset1           offset in first sequence
&nbsp;     * @param length1           length of first sequence&#39;s part to be aligned including artificially added letters
&nbsp;     * @param addedNucleotides1 number of artificially added letters to the first sequence
&nbsp;     * @param offset2           offset in second sequence
&nbsp;     * @param length2           length of second sequence&#39;s part to be aligned including artificially added letters
&nbsp;     * @param addedNucleotides2 number of artificially added letters to the second sequence
&nbsp;     * @param width             width of banded alignment matrix. In other terms max allowed number of indels
&nbsp;     */
&nbsp;    public static Alignment&lt;NucleotideSequence&gt; alignRightAdded(LinearGapAlignmentScoring scoring, NucleotideSequence seq1, NucleotideSequence seq2,
&nbsp;                                                                int offset1, int length1, int addedNucleotides1, int offset2, int length2, int addedNucleotides2,
&nbsp;                                                                int width) {
&nbsp;        try {
<b class="nc">&nbsp;            MutationsBuilder&lt;NucleotideSequence&gt; mutations = new MutationsBuilder&lt;&gt;(NucleotideSequence.ALPHABET);</b>
<b class="nc">&nbsp;            BandedSemiLocalResult result = alignRightAdded0(scoring, seq1, seq2,</b>
&nbsp;                    offset1, length1, addedNucleotides1, offset2, length2, addedNucleotides2,
<b class="nc">&nbsp;                    width, mutations, AlignmentCache.get());</b>
<b class="nc">&nbsp;            return new Alignment&lt;&gt;(seq1, mutations.createAndDestroy(),</b>
&nbsp;                    new Range(offset1, result.sequence1Stop + 1), new Range(offset2, result.sequence2Stop + 1),
&nbsp;                    result.score);
&nbsp;        } finally {
<b class="nc">&nbsp;            AlignmentCache.release();</b>
<b class="nc">&nbsp;        }</b>
&nbsp;    }
&nbsp;
&nbsp;    /**
&nbsp;     * Alignment which identifies what is the highly similar part of the both sequences.
&nbsp;     * &lt;p/&gt;
&nbsp;     * &lt;p&gt;Alignment is done in the way that beginning of second sequences is aligned to beginning of first
&nbsp;     * sequence.&lt;/p&gt;
&nbsp;     * &lt;p/&gt;
&nbsp;     * &lt;p&gt;Alignment terminates when score in banded alignment matrix reaches {@code stopPenalty} value.&lt;/p&gt;
&nbsp;     * &lt;p/&gt;
&nbsp;     * &lt;p&gt;In other words, only left part of second sequence is to be aligned&lt;/p&gt;
&nbsp;     *
&nbsp;     * @param scoring     scoring system
&nbsp;     * @param seq1        first sequence
&nbsp;     * @param seq2        second sequence
&nbsp;     * @param offset1     offset in first sequence
&nbsp;     * @param length1     length of first sequence&#39;s part to be aligned
&nbsp;     * @param offset2     offset in second sequence
&nbsp;     * @param length2     length of second sequence&#39;s part to be aligned@param width
&nbsp;     * @param stopPenalty alignment score value in banded alignment matrix at which alignment terminates
&nbsp;     * @return object which contains positions at which alignment terminated and array of mutations
&nbsp;     */
&nbsp;    public static Alignment&lt;NucleotideSequence&gt; alignSemiLocalLeft(LinearGapAlignmentScoring scoring, NucleotideSequence seq1, NucleotideSequence seq2,
&nbsp;                                                                   int offset1, int length1, int offset2, int length2,
&nbsp;                                                                   int width, int stopPenalty) {
&nbsp;        try {
<b class="nc">&nbsp;            int minLength = Math.min(length1, length2) + width + 1;</b>
<b class="nc">&nbsp;            length1 = Math.min(length1, minLength);</b>
<b class="nc">&nbsp;            length2 = Math.min(length2, minLength);</b>
<b class="nc">&nbsp;            MutationsBuilder&lt;NucleotideSequence&gt; mutations = new MutationsBuilder&lt;&gt;(NucleotideSequence.ALPHABET);</b>
<b class="nc">&nbsp;            BandedSemiLocalResult result = alignSemiLocalLeft0(scoring, seq1, seq2,</b>
<b class="nc">&nbsp;                    offset1, length1, offset2, length2, width, stopPenalty, mutations, AlignmentCache.get());</b>
<b class="nc">&nbsp;            return new Alignment&lt;&gt;(seq1, mutations.createAndDestroy(),</b>
&nbsp;                    new Range(offset1, result.sequence1Stop + 1), new Range(offset2, result.sequence2Stop + 1),
&nbsp;                    result.score);
&nbsp;        } finally {
<b class="nc">&nbsp;            AlignmentCache.release();</b>
<b class="nc">&nbsp;        }</b>
&nbsp;    }
&nbsp;
&nbsp;    /**
&nbsp;     * Alignment which identifies what is the highly similar part of the both sequences.
&nbsp;     * &lt;p/&gt;
&nbsp;     * &lt;p&gt;Alignment is done in the way that beginning of second sequences is aligned to beginning of first
&nbsp;     * sequence.&lt;/p&gt;
&nbsp;     * &lt;p/&gt;
&nbsp;     * &lt;p&gt;Alignment terminates when score in banded alignment matrix reaches {@code stopPenalty} value.&lt;/p&gt;
&nbsp;     * &lt;p/&gt;
&nbsp;     * &lt;p&gt;In other words, only left part of second sequence is to be aligned&lt;/p&gt;
&nbsp;     *
&nbsp;     * @param scoring     scoring system
&nbsp;     * @param seq1        first sequence
&nbsp;     * @param seq2        second sequence
&nbsp;     * @param stopPenalty alignment score value in banded alignment matrix at which alignment terminates
&nbsp;     * @return object which contains positions at which alignment terminated and array of mutations
&nbsp;     */
&nbsp;    public static Alignment&lt;NucleotideSequence&gt; alignSemiLocalLeft(LinearGapAlignmentScoring scoring,
&nbsp;                                                                   NucleotideSequence seq1, NucleotideSequence seq2,
&nbsp;                                                                   int width, int stopPenalty) {
<b class="nc">&nbsp;        return alignSemiLocalLeft(scoring, seq1, seq2, 0, seq1.size(), 0, seq2.size(), width, stopPenalty);</b>
&nbsp;    }
&nbsp;
&nbsp;    /**
&nbsp;     * Alignment which identifies what is the highly similar part of the both sequences.
&nbsp;     * &lt;p/&gt;
&nbsp;     * &lt;p&gt;Alignment is done in the way that end of second sequence is aligned to end of first sequence.&lt;/p&gt; &lt;p&gt;Alignment
&nbsp;     * terminates when score in banded alignment matrix reaches {@code stopPenalty} value.&lt;/p&gt; &lt;p&gt;In other words, only
&nbsp;     * right part of second sequence is to be aligned.&lt;/p&gt;
&nbsp;     *
&nbsp;     * @param scoring     scoring system
&nbsp;     * @param seq1        first sequence
&nbsp;     * @param seq2        second sequence
&nbsp;     * @param offset1     offset in first sequence
&nbsp;     * @param length1     length of first sequence&#39;s part to be aligned
&nbsp;     * @param offset2     offset in second sequence
&nbsp;     * @param length2     length of second sequence&#39;s part to be aligned@param width
&nbsp;     * @param stopPenalty alignment score value in banded alignment matrix at which alignment terminates
&nbsp;     * @return object which contains positions at which alignment terminated and array of mutations
&nbsp;     */
&nbsp;    public static Alignment&lt;NucleotideSequence&gt; alignSemiLocalRight(LinearGapAlignmentScoring scoring, NucleotideSequence seq1, NucleotideSequence seq2,
&nbsp;                                                                    int offset1, int length1, int offset2, int length2,
&nbsp;                                                                    int width, int stopPenalty) {
&nbsp;        try {
<b class="nc">&nbsp;            int minLength = Math.min(length1, length2) + width + 1;</b>
<b class="nc">&nbsp;            int l1 = Math.min(length1, minLength);</b>
<b class="nc">&nbsp;            int l2 = Math.min(length2, minLength);</b>
<b class="nc">&nbsp;            offset1 = offset1 + length1 - l1;</b>
<b class="nc">&nbsp;            offset2 = offset2 + length2 - l2;</b>
<b class="nc">&nbsp;            length1 = l1;</b>
<b class="nc">&nbsp;            length2 = l2;</b>
<b class="nc">&nbsp;            MutationsBuilder&lt;NucleotideSequence&gt; mutations = new MutationsBuilder&lt;&gt;(NucleotideSequence.ALPHABET);</b>
<b class="nc">&nbsp;            BandedSemiLocalResult result = alignSemiLocalRight0(scoring, seq1, seq2,</b>
&nbsp;                    offset1, length1, offset2, length2, width,
<b class="nc">&nbsp;                    stopPenalty, mutations, AlignmentCache.get());</b>
<b class="nc">&nbsp;            return new Alignment&lt;&gt;(seq1, mutations.createAndDestroy(),</b>
&nbsp;                    new Range(result.sequence1Stop, offset1 + length1), new Range(result.sequence2Stop, offset2 + length2),
&nbsp;                    result.score);
&nbsp;        } finally {
<b class="nc">&nbsp;            AlignmentCache.release();</b>
<b class="nc">&nbsp;        }</b>
&nbsp;    }
&nbsp;
&nbsp;    /**
&nbsp;     * Alignment which identifies what is the highly similar part of the both sequences.
&nbsp;     * &lt;p/&gt;
&nbsp;     * &lt;p&gt;Alignment is done in the way that end of second sequence is aligned to end of first sequence.&lt;/p&gt; &lt;p&gt;Alignment
&nbsp;     * terminates when score in banded alignment matrix reaches {@code stopPenalty} value.&lt;/p&gt; &lt;p&gt;In other words, only
&nbsp;     * right part of second sequence is to be aligned.&lt;/p&gt;
&nbsp;     *
&nbsp;     * @param scoring     scoring system
&nbsp;     * @param seq1        first sequence
&nbsp;     * @param seq2        second sequence
&nbsp;     * @param stopPenalty alignment score value in banded alignment matrix at which alignment terminates
&nbsp;     * @return object which contains positions at which alignment terminated and array of mutations
&nbsp;     */
&nbsp;    public static Alignment&lt;NucleotideSequence&gt; alignSemiLocalRight(LinearGapAlignmentScoring scoring,
&nbsp;                                                                    NucleotideSequence seq1, NucleotideSequence seq2,
&nbsp;                                                                    int width, int stopPenalty) {
<b class="nc">&nbsp;        return alignSemiLocalRight(scoring, seq1, seq2, 0, seq1.size(), 0, seq2.size(), width, stopPenalty);</b>
&nbsp;    }
&nbsp;}
</code>
</pre>
</div>

<script type="text/javascript">
(function() {
    var msie = false, msie9 = false;
    /*@cc_on
      msie = true;
      @if (@_jscript_version >= 9)
        msie9 = true;
      @end
    @*/

    if (!msie || msie && msie9) {
      hljs.highlightAll()
      hljs.initLineNumbersOnLoad();
    }
})();
</script>

<div class="footer">
    
    <div style="float:right;">generated on 2024-04-04 10:47</div>
</div>
</body>
</html>
