


<!DOCTYPE html>
<html id="htmlId">
<head>
  <meta http-equiv="Content-Type" content="text/html;charset=UTF-8"> 
  <title>Coverage Report > MismatchOnlyPairedReadMerger</title>
  <style type="text/css">
    @import "../../css/coverage.css";
    @import "../../css/idea.min.css";
  </style>
  <script type="text/javascript" src="../../js/highlight.min.js"></script>
  <script type="text/javascript" src="../../js/highlightjs-line-numbers.min.js"></script>
</head>

<body>
<div class="content">
<div class="breadCrumbs">
Current scope:     <a href="../../index.html">all classes</a>
    <span class="separator">|</span>
    <a href="../index.html">com.milaboratory.core.merger</a>
</div>

<h1>Coverage Summary for Class: MismatchOnlyPairedReadMerger (com.milaboratory.core.merger)</h1>

<table class="coverageStats">

<tr>
  <th class="name">Class</th>
<th class="coverageStat 
">
  Method, %
</th>
<th class="coverageStat 
">
  Line, %
</th>
</tr>
<tr>
  <td class="name">MismatchOnlyPairedReadMerger</td>
<td class="coverageStat">
  <span class="percent">
    0%
  </span>
  <span class="absValue">
    (0/9)
  </span>
</td>
<td class="coverageStat">
  <span class="percent">
    0%
  </span>
  <span class="absValue">
    (0/125)
  </span>
</td>
</tr>
  <tr>
    <td class="name">MismatchOnlyPairedReadMerger$1</td>
<td class="coverageStat">
  <span class="percent">
    0%
  </span>
  <span class="absValue">
    (0/1)
  </span>
</td>
<td class="coverageStat">
  <span class="percent">
    0%
  </span>
  <span class="absValue">
    (0/2)
  </span>
</td>
  </tr>
<tr>
  <td class="name"><strong>Total</strong></td>
<td class="coverageStat">
  <span class="percent">
    0%
  </span>
  <span class="absValue">
    (0/10)
  </span>
</td>
<td class="coverageStat">
  <span class="percent">
    0%
  </span>
  <span class="absValue">
    (0/127)
  </span>
</td>
</tr>
</table>

<br/>
<br/>


<pre>
<code class="sourceCode" id="sourceCode">&nbsp;/*
&nbsp; * Copyright 2016 MiLaboratory.com
&nbsp; *
&nbsp; * Licensed under the Apache License, Version 2.0 (the &quot;License&quot;);
&nbsp; * you may not use this file except in compliance with the License.
&nbsp; * You may obtain a copy of the License at
&nbsp; *
&nbsp; *     http://www.apache.org/licenses/LICENSE-2.0
&nbsp; *
&nbsp; * Unless required by applicable law or agreed to in writing, software
&nbsp; * distributed under the License is distributed on an &quot;AS IS&quot; BASIS,
&nbsp; * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
&nbsp; * See the License for the specific language governing permissions and
&nbsp; * limitations under the License.
&nbsp; */
&nbsp;package com.milaboratory.core.merger;
&nbsp;
&nbsp;import cc.redberry.pipe.Processor;
&nbsp;import com.milaboratory.core.PairedEndReadsLayout;
&nbsp;import com.milaboratory.core.io.sequence.PairedRead;
&nbsp;import com.milaboratory.core.merger.MergerParameters.IdentityType;
&nbsp;import com.milaboratory.core.motif.BitapMatcher;
&nbsp;import com.milaboratory.core.motif.BitapPattern;
&nbsp;import com.milaboratory.core.motif.Motif;
&nbsp;import com.milaboratory.core.motif.MotifUtils;
&nbsp;import com.milaboratory.core.sequence.NSequenceWithQuality;
&nbsp;import com.milaboratory.core.sequence.NucleotideSequence;
&nbsp;import com.milaboratory.core.sequence.SequenceBuilder;
&nbsp;import com.milaboratory.core.sequence.SequenceQualityBuilder;
&nbsp;
&nbsp;import static com.milaboratory.core.sequence.SequencesUtils.mismatchCount;
&nbsp;import static java.lang.Math.*;
&nbsp;
<b class="nc">&nbsp;public final class MismatchOnlyPairedReadMerger implements Processor&lt;PairedRead, PairedReadMergingResult&gt;,</b>
&nbsp;                                                           java.io.Serializable {
&nbsp;    public static final int MIN_SCORE_VALUE = 0;
&nbsp;    final int minOverlap;
&nbsp;    final double minimalIdentity;
&nbsp;    final IdentityType identityType;
&nbsp;    final double maxMismatchesPart;
&nbsp;    final int maxScoreValue;
&nbsp;    // opposite reads direction or collinear
&nbsp;    final boolean[] strands;
&nbsp;    final int motifLength;
&nbsp;    final int maxMismatchesInMotif;
&nbsp;    final QualityMergingAlgorithm qualityMergingAlgorithm;
&nbsp;    final PairedEndReadsLayout pairedEndReadsLayout;
&nbsp;
&nbsp;    /**
&nbsp;     * Creates paired-end reads merger for Illumina (or other opposite reads) data.
&nbsp;     *
&nbsp;     * @param parameters merger parameters
&nbsp;     */
&nbsp;    public MismatchOnlyPairedReadMerger(MergerParameters parameters) {
<b class="nc">&nbsp;        this(parameters.getMinimalOverlap(), parameters.getMinimalIdentity(), parameters.getIdentityType(), parameters.getMaxQuality(),</b>
<b class="nc">&nbsp;                parameters.qualityMergingAlgorithm, parameters.getPartsLayout());</b>
&nbsp;    }
&nbsp;
&nbsp;    /**
&nbsp;     * Creates paired-end reads merger.
&nbsp;     *
&nbsp;     * @param minOverlap              minimal number of nucleotide in overlap region
&nbsp;     * @param minimalIdentity         maximal allowed percent of mismatches in overlap region
&nbsp;     * @param maxScoreValue           maximal output quality score value
&nbsp;     * @param qualityMergingAlgorithm algorithm to infer quality of merged reads from it&#39;s pairs
&nbsp;     * @param pairedEndReadsLayout    orientation of read pairs
&nbsp;     */
&nbsp;    public MismatchOnlyPairedReadMerger(int minOverlap, double minimalIdentity,
&nbsp;                                        int maxScoreValue,
&nbsp;                                        QualityMergingAlgorithm qualityMergingAlgorithm,
&nbsp;                                        PairedEndReadsLayout pairedEndReadsLayout) {
<b class="nc">&nbsp;        this(minOverlap, minimalIdentity, IdentityType.Unweighted, maxScoreValue, qualityMergingAlgorithm, pairedEndReadsLayout);</b>
&nbsp;    }
&nbsp;
&nbsp;    /**
&nbsp;     * Creates paired-end reads merger.
&nbsp;     *
&nbsp;     * @param minOverlap              minimal number of nucleotide in overlap region
&nbsp;     * @param minimalIdentity         maximal allowed percent of mismatches in overlap region
&nbsp;     * @param identityType            identity type
&nbsp;     * @param maxScoreValue           maximal output quality score value
&nbsp;     * @param qualityMergingAlgorithm algorithm to infer quality of merged reads from it&#39;s pairs
&nbsp;     * @param pairedEndReadsLayout    orientation of read pairs
&nbsp;     */
&nbsp;    public MismatchOnlyPairedReadMerger(int minOverlap, double minimalIdentity, IdentityType identityType,
&nbsp;                                        int maxScoreValue,
&nbsp;                                        QualityMergingAlgorithm qualityMergingAlgorithm,
<b class="nc">&nbsp;                                        PairedEndReadsLayout pairedEndReadsLayout) {</b>
<b class="nc">&nbsp;        if (qualityMergingAlgorithm == null || pairedEndReadsLayout == null)</b>
<b class="nc">&nbsp;            throw new NullPointerException();</b>
<b class="nc">&nbsp;        this.qualityMergingAlgorithm = qualityMergingAlgorithm;</b>
<b class="nc">&nbsp;        this.pairedEndReadsLayout = pairedEndReadsLayout;</b>
<b class="nc">&nbsp;        this.minOverlap = minOverlap;</b>
<b class="nc">&nbsp;        this.minimalIdentity = minimalIdentity;</b>
<b class="nc">&nbsp;        this.identityType = identityType;</b>
<b class="nc">&nbsp;        this.maxMismatchesPart = 1.0 - minimalIdentity;</b>
<b class="nc">&nbsp;        this.strands = pairedEndReadsLayout.getPossibleRelativeStrands();</b>
<b class="nc">&nbsp;        this.maxScoreValue = maxScoreValue;</b>
&nbsp;
&nbsp;        // Calculating length fo motif to be used in Bitap search.
<b class="nc">&nbsp;        this.motifLength = min(minOverlap, 62);</b>
<b class="nc">&nbsp;        this.maxMismatchesInMotif = (int) round(motifLength * maxMismatchesPart);</b>
&nbsp;    }
&nbsp;
&nbsp;    public PairedReadMergingResult merge(NSequenceWithQuality read1p, NSequenceWithQuality read2p) {
<b class="nc">&nbsp;        return merge(read1p, read2p, null);</b>
&nbsp;    }
&nbsp;
&nbsp;    public PairedReadMergingResult merge(NSequenceWithQuality read1p, NSequenceWithQuality read2p,
&nbsp;                                         PairedRead pairedRead) {
&nbsp;        // If there is no sufficient letters in one of read overlapping is impossible
<b class="nc">&nbsp;        if (read1p.size() &lt; minOverlap || read2p.size() &lt; minOverlap)</b>
&nbsp;            // Return failed result
<b class="nc">&nbsp;            return new PairedReadMergingResult(pairedRead);</b>
&nbsp;
<b class="nc">&nbsp;        PairedReadMergingResult ret = null;</b>
&nbsp;
<b class="nc">&nbsp;        for (boolean strand : strands) {</b>
<b class="nc">&nbsp;            NSequenceWithQuality read1 = read1p;</b>
&nbsp;
&nbsp;            // Making reverse complement from second read to bring reads to the same strand
&nbsp;            // (if reads configuration is opposite)
<b class="nc">&nbsp;            NSequenceWithQuality read2 = strand ? read2p.getReverseComplement() : read2p;</b>
&nbsp;
<b class="nc">&nbsp;            boolean swapped = false;</b>
&nbsp;            // read2 always smaller then read1
<b class="nc">&nbsp;            if (read2.size() &gt; read1.size()) {</b>
<b class="nc">&nbsp;                NSequenceWithQuality tmp = read1;</b>
<b class="nc">&nbsp;                read1 = read2;</b>
<b class="nc">&nbsp;                read2 = tmp;</b>
<b class="nc">&nbsp;                swapped = true;</b>
&nbsp;            }
&nbsp;
&nbsp;            // Searching
&nbsp;
&nbsp;            // Creating bitap pattern for beginning and ending of read2
<b class="nc">&nbsp;            Motif&lt;NucleotideSequence&gt; motif = MotifUtils.twoSequenceMotif(</b>
<b class="nc">&nbsp;                    read2.getSequence(), 0,</b>
<b class="nc">&nbsp;                    read2.getSequence(), read2.size() - motifLength,</b>
&nbsp;                    motifLength
&nbsp;            );
<b class="nc">&nbsp;            BitapPattern bitapPattern = motif.getBitapPattern();</b>
<b class="nc">&nbsp;            BitapMatcher bitapMatcher = bitapPattern.substitutionOnlyMatcherFirst(maxMismatchesInMotif, read1.getSequence());</b>
&nbsp;
&nbsp;            int matchPosition, mismatches, overlap;
&nbsp;            double identity;
&nbsp;
<b class="nc">&nbsp;            PairedReadMergingResult tmp = null;</b>
<b class="nc">&nbsp;            while ((matchPosition = bitapMatcher.findNext()) != -1) {</b>
&nbsp;
&nbsp;                // Case: beginning of r2 matched
&nbsp;
&nbsp;                // Finally checking current hit position
<b class="nc">&nbsp;                overlap = min(read1.size() - matchPosition, read2.size());</b>
<b class="nc">&nbsp;                mismatches = mismatchCount(</b>
<b class="nc">&nbsp;                        read1.getSequence(), matchPosition,</b>
<b class="nc">&nbsp;                        read2.getSequence(), 0,</b>
&nbsp;                        overlap);
<b class="nc">&nbsp;                identity = identity(identityType,</b>
&nbsp;                        read1, matchPosition,
&nbsp;                        read2, 0,
&nbsp;                        overlap);
<b class="nc">&nbsp;                if (identity &gt;= minimalIdentity) {</b>
<b class="nc">&nbsp;                    tmp = new PairedReadMergingResult(pairedRead, overlap(read1, read2, matchPosition),</b>
&nbsp;                            overlap, mismatches, strand, swapped ? -matchPosition : matchPosition, identityType, identity);
<b class="nc">&nbsp;                    break;</b>
&nbsp;                }
&nbsp;
&nbsp;                // Case: ending of r2 matched
<b class="nc">&nbsp;                matchPosition += motifLength; // Calculating position of right overlap boundary</b>
<b class="nc">&nbsp;                overlap = min(matchPosition, read2.size());</b>
&nbsp;
<b class="nc">&nbsp;                mismatches = mismatchCount(</b>
<b class="nc">&nbsp;                        read1.getSequence(), matchPosition - overlap,</b>
<b class="nc">&nbsp;                        read2.getSequence(), max(0, read2.size() - overlap),</b>
&nbsp;                        overlap);
<b class="nc">&nbsp;                identity = identity(identityType,</b>
&nbsp;                        read1, matchPosition - overlap,
<b class="nc">&nbsp;                        read2, max(0, read2.size() - overlap),</b>
&nbsp;                        overlap);
<b class="nc">&nbsp;                if (identity &gt;= minimalIdentity) {</b>
<b class="nc">&nbsp;                    final int offset = min(matchPosition - read2.size(), 0);</b>
<b class="nc">&nbsp;                    tmp = new PairedReadMergingResult(pairedRead, overlap(read1, read2, offset),</b>
&nbsp;                            overlap, mismatches, strand, swapped ? -offset : offset, identityType, identity);
<b class="nc">&nbsp;                    break;</b>
&nbsp;                }
&nbsp;            }
<b class="nc">&nbsp;            if (tmp != null &amp;&amp; (ret == null || ret.score() &lt; tmp.score()))</b>
<b class="nc">&nbsp;                ret = tmp;</b>
&nbsp;        }
&nbsp;
<b class="nc">&nbsp;        if (ret == null)</b>
<b class="nc">&nbsp;            return new PairedReadMergingResult(pairedRead);</b>
&nbsp;        else
<b class="nc">&nbsp;            return ret;</b>
&nbsp;    }
&nbsp;
&nbsp;    public static double identity(IdentityType identityType,
&nbsp;                                  NSequenceWithQuality seq1, int offset1,
&nbsp;                                  NSequenceWithQuality seq2, int offset2,
&nbsp;                                  int length) {
<b class="nc">&nbsp;        if (length == 0)</b>
<b class="nc">&nbsp;            return 0.0;</b>
<b class="nc">&nbsp;        NucleotideSequence seq1s = seq1.getSequence(), seq2s = seq2.getSequence();</b>
<b class="nc">&nbsp;        switch (identityType) {</b>
&nbsp;            case Unweighted:
<b class="nc">&nbsp;                return 1.0 * (length - mismatchCount(seq1s, offset1, seq2s, offset2, length)) / length;</b>
&nbsp;            case MinimalQualityWeighted:
<b class="nc">&nbsp;                if (seq1.size() &lt; offset1 + length || seq2.size() &lt; offset2 + length)</b>
<b class="nc">&nbsp;                    throw new IllegalArgumentException();</b>
&nbsp;
<b class="nc">&nbsp;                long identQuality = 0, totalQuality = 0;</b>
<b class="nc">&nbsp;                int nIdentical = 0;</b>
<b class="nc">&nbsp;                for (int i = 0; i &lt; length; ++i) {</b>
<b class="nc">&nbsp;                    int minQuality = Math.min(</b>
<b class="nc">&nbsp;                            seq1.getQuality().value(i + offset1),</b>
<b class="nc">&nbsp;                            seq2.getQuality().value(i + offset2));</b>
<b class="nc">&nbsp;                    if (seq1s.codeAt(i + offset1) == seq2s.codeAt(i + offset2)) {</b>
<b class="nc">&nbsp;                        identQuality += minQuality;</b>
<b class="nc">&nbsp;                        ++nIdentical;</b>
&nbsp;                    }
<b class="nc">&nbsp;                    totalQuality += minQuality;</b>
&nbsp;                }
<b class="nc">&nbsp;                if (totalQuality == 0)</b>
<b class="nc">&nbsp;                    return 1.0 * nIdentical / length;</b>
<b class="nc">&nbsp;                return 1.0 * identQuality / totalQuality;</b>
&nbsp;
&nbsp;            default:
<b class="nc">&nbsp;                throw new RuntimeException(&quot;not supported identity type: &quot; + identityType);</b>
&nbsp;        }
&nbsp;    }
&nbsp;
&nbsp;    @Override
&nbsp;    public PairedReadMergingResult process(PairedRead pairedRead) {
<b class="nc">&nbsp;        NSequenceWithQuality read1p = pairedRead.getR1().getData();</b>
<b class="nc">&nbsp;        NSequenceWithQuality read2p = pairedRead.getR2().getData();</b>
<b class="nc">&nbsp;        return merge(read1p, read2p, pairedRead);</b>
&nbsp;    }
&nbsp;
&nbsp;    /**
&nbsp;     * @param seq1   sequence 1
&nbsp;     * @param seq2   sequence 2
&nbsp;     * @param offset position of first nucleotide of seq2 in seq1
&nbsp;     * @return overlapped sequence
&nbsp;     */
&nbsp;    public NSequenceWithQuality overlap(NSequenceWithQuality seq1, NSequenceWithQuality seq2, int offset) {
<b class="nc">&nbsp;        if (qualityMergingAlgorithm == null)</b>
<b class="nc">&nbsp;            throw new NullPointerException();</b>
&nbsp;
&nbsp;        // Calculating length of resulting sequence
<b class="nc">&nbsp;        int length = abs(offset) +</b>
&nbsp;                (offset &gt;= 0 ?
<b class="nc">&nbsp;                        max(seq1.size() - offset, seq2.size()) :</b>
<b class="nc">&nbsp;                        max(seq1.size(), seq2.size() + offset) // offset is negative here</b>
&nbsp;                );
&nbsp;
<b class="nc">&nbsp;        SequenceBuilder&lt;NucleotideSequence&gt; seqBuilder = NucleotideSequence.ALPHABET.createBuilder()</b>
<b class="nc">&nbsp;                .ensureCapacity(length);</b>
<b class="nc">&nbsp;        SequenceQualityBuilder qualBuilder = new SequenceQualityBuilder().ensureCapacity(length);</b>
&nbsp;
&nbsp;        byte quality, letter, l, q;
<b class="nc">&nbsp;        int from = min(0, offset);</b>
<b class="nc">&nbsp;        int position, to = length + from;</b>
<b class="nc">&nbsp;        for (int i = from; i &lt; to; ++i) {</b>
<b class="nc">&nbsp;            quality = 0;</b>
<b class="nc">&nbsp;            letter = -1;</b>
&nbsp;
&nbsp;            // Checking read 1
<b class="nc">&nbsp;            if (i &gt;= 0 &amp;&amp; i &lt; seq1.size()) {</b>
<b class="nc">&nbsp;                quality = seq1.getQuality().value(i);</b>
<b class="nc">&nbsp;                letter = seq1.getSequence().codeAt(i);</b>
&nbsp;            }
&nbsp;
&nbsp;            // Checking read 2
<b class="nc">&nbsp;            position = i - offset;</b>
<b class="nc">&nbsp;            if (position &gt;= 0 &amp;&amp; position &lt; seq2.size()) {</b>
<b class="nc">&nbsp;                l = seq2.getSequence().codeAt(position);</b>
<b class="nc">&nbsp;                q = seq2.getQuality().value(position);</b>
<b class="nc">&nbsp;                if (letter == -1) { // Letter not initialized</b>
<b class="nc">&nbsp;                    letter = l;</b>
<b class="nc">&nbsp;                    quality = q;</b>
<b class="nc">&nbsp;                } else if (letter == l) // Same letter in both reads</b>
<b class="nc">&nbsp;                    switch (qualityMergingAlgorithm) {</b>
&nbsp;                        case SumSubtraction:
&nbsp;                        case SumMax:
<b class="nc">&nbsp;                            quality = (byte) min(maxScoreValue, quality + q);</b>
&nbsp;                        case MaxSubtraction:
&nbsp;                        case MaxMax:
<b class="nc">&nbsp;                            quality = (byte) max(quality, q);</b>
&nbsp;                    }
&nbsp;                else // Different letter
<b class="nc">&nbsp;                    switch (qualityMergingAlgorithm) {</b>
&nbsp;                        case MaxSubtraction:
&nbsp;                        case SumSubtraction:
<b class="nc">&nbsp;                            if (q &gt; quality) {</b>
<b class="nc">&nbsp;                                letter = l;</b>
<b class="nc">&nbsp;                                quality = (byte) max(MIN_SCORE_VALUE, q - quality);</b>
&nbsp;                            } else
<b class="nc">&nbsp;                                quality = (byte) max(MIN_SCORE_VALUE, quality - q);</b>
<b class="nc">&nbsp;                            break;</b>
&nbsp;                        case MaxMax:
&nbsp;                        case SumMax:
<b class="nc">&nbsp;                            if (q &gt; quality) {</b>
<b class="nc">&nbsp;                                letter = l;</b>
<b class="nc">&nbsp;                                quality = q;</b>
&nbsp;                            }
&nbsp;                            break;
&nbsp;                    }
&nbsp;            }
&nbsp;
<b class="nc">&nbsp;            assert letter != -1;</b>
&nbsp;
<b class="nc">&nbsp;            seqBuilder.append(letter);</b>
<b class="nc">&nbsp;            qualBuilder.append(quality);</b>
&nbsp;        }
&nbsp;
<b class="nc">&nbsp;        return new NSequenceWithQuality(seqBuilder.createAndDestroy(), qualBuilder.createAndDestroy());</b>
&nbsp;    }
&nbsp;}
</code>
</pre>
</div>

<script type="text/javascript">
(function() {
    var msie = false, msie9 = false;
    /*@cc_on
      msie = true;
      @if (@_jscript_version >= 9)
        msie9 = true;
      @end
    @*/

    if (!msie || msie && msie9) {
      hljs.highlightAll()
      hljs.initLineNumbersOnLoad();
    }
})();
</script>

<div class="footer">
    
    <div style="float:right;">generated on 2024-04-04 10:47</div>
</div>
</body>
</html>
